feat: allow patching of resource class

olevski opened the pull request:
This allows the resource class on a running server to be patched. But only if the server is hibernated.

The reason for the limitation is to avoid people from inadvertantly restarting their sessions. Because if we patch the resource class when the server is running k8s will restart the session pod.

/deploy renku-ui=leafty/2941-feat-modify-session renku=rclone-sessions extra-values=amalthea.image.repository=registry.dev.renku.ch/tasko.olevski/renku-images/amalthea,amalthea.image.tag=0.10.0-0.dev.git.205.hb1d4c56

leafty commented on renku_notebooks/api/schemas/servers_patch.py:
```diff
@@ -17,4 +17,5 @@ def list(cls):
 
-    state = fields.String(required=True, validate=validate.OneOf(PatchServerStatusEnum.list()))
+    state = fields.String(required=False, validate=validate.OneOf(PatchServerStatusEnum.list()))
+    resource_class_id = fields.Int(required=False, validate=lambda x: x > 0)
```
The `resource_class_id` is not part of the Session object given by the notebooks API. Would it be possible to store it somewhere (annotations?) so that the UI can know which is the current resource class?

olevski replied:
Ok so the resource class is now in the annotations like `"renku.io/resourceClassId": "2"`.

However keep in mind that this field may be completely absent in the response because we never recorded the resource class in the JupyterServer resource up until this PR.

So the UI should handle this. I have no way to figure out what resource class was used to start the session on the sessions that are already there or launched prior to this PR.

leafty replied:
👍

leafty commented on renku_notebooks/api/notebooks.py:
```diff
@@ -438,13 +438,43 @@ def patch_server(user, server_name, state):
+        parsed_server_options = config.crc_validator.validate_class_storage(
+            user, resource_class_id, storage=None  # we do not care about validating storage
+        )
+        patch = [
```
Here, patching the k8s quota is missing.

olevski commented on renku_notebooks/api/notebooks.py:
```diff
@@ -438,13 +438,68 @@ def patch_server(user, server_name, state):
 
     server = config.k8s.client.get_server(server_name, user.safe_username)
+    new_server = server
+    currently_hibernated = server.get("spec", {}).get("jupyterServer", {}).get("hibernated", False)
```
I modified the code so that users are allowed to patch failing sessions.

leafty replied:
I sometimes get this error:
```
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server rejected our request due to an error in our request","reason":"Invalid","details":{},"code":422}

Traceback (most recent call last):
  File "/app/renku_notebooks/api/classes/k8s_client.py", line 132, in patch_server
    server = self._custom_objects_patch.patch_namespaced_custom_object(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/api/custom_objects_api.py", line 2941, in patch_namespaced_custom_object
    return self.patch_namespaced_custom_object_with_http_info(group, version, namespace, plural, name, body, **kwargs)  # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/api/custom_objects_api.py", line 3072, in patch_namespaced_custom_object_with_http_info
    return self.api_client.call_api(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
                    ^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/api_client.py", line 407, in request
    return self.rest_client.PATCH(url,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/rest.py", line 299, in PATCH
    return self.request("PATCH", url,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.11/site-packages/kubernetes/client/rest.py", line 238, in request
    raise ApiException(http_resp=r)
kubernetes.client.exceptions.ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Audit-Id': '31a6f292-8768-44bb-af50-f50252d14806', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '644b2a2f-13de-45a7-a037-0c1722afb9df', 'X-Kubernetes-Pf-Prioritylevel-Uid': '5c054d07-27dd-4912-9eca-f523ae31544a', 'Date': 'Thu, 11 Jan 2024 10:50:17 GMT', 'Content-Length': '187'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server rejected our request due to an error in our request","reason":"Invalid","details":{},"code":422}
```

leafty commented on renku_notebooks/api/notebooks.py:
```diff
@@ -438,13 +438,71 @@ def patch_server(user, server_name, state):
+                {
+                    "op": "remove",
+                    # NOTE: ~1 is how you escape '/' in json-patch
+                    "path": "/metadata/labels/renku.io~1quota",
```
it seems that `"value"` needs to be removed here too

leafty commented:
I am also seeing a case where the server got running but is still reported as failed:
```json
{
        "flora-2eth-another-2dplayground-2dp-465de908": {
            "annotations": {
                "renku.io/branch": "master",
                "renku.io/commit-sha": "3869e389a4353455d82801b4806b04cbb43373d8",
                "renku.io/default_image_used": "False",
                "renku.io/git-host": "gitlab.dev.renku.ch",
                "renku.io/gitlabProjectId": "100761",
                "renku.io/hibernatedSecondsThreshold": "259200",
                "renku.io/hibernation": "",
                "renku.io/hibernationBranch": "",
                "renku.io/hibernationCommitSha": "",
                "renku.io/hibernationDate": "",
                "renku.io/hibernationDirty": "",
                "renku.io/hibernationSynchronized": "",
                "renku.io/idleSecondsThreshold": "86400",
                "renku.io/lastActivityDate": "",
                "renku.io/namespace": "johann.thiebaut1",
                "renku.io/projectName": "another-playground-project",
                "renku.io/repository": "https://gitlab.dev.renku.ch/johann.thiebaut1/another-playground-project",
                "renku.io/resourceClassId": "2",
                "renku.io/servername": "flora-2eth-another-2dplayground-2dp-465de908",
                "renku.io/username": "flora.thiebaut@sdsc.ethz.ch"
            },
            "cloudstorage": [],
            "image": "registry.dev.renku.ch/johann.thiebaut1/another-playground-project:3869e38",
            "name": "flora-2eth-another-2dplayground-2dp-465de908",
            "resources": {
                "requests": {
                    "cpu": 0.5,
                    "memory": "2G",
                    "storage": "4G"
                },
                "usage": {}
            },
            "started": "2024-01-11T10:45:17+00:00",
            "state": {
                "pod_name": "flora-2eth-another-2dplayground-2dp-465de908-0"
            },
            "status": {
                "details": [
                    {
                        "status": "ready",
                        "step": "Initialization"
                    },
                    {
                        "status": "ready",
                        "step": "Downloading server image"
                    },
                    {
                        "status": "ready",
                        "step": "Cloning and configuring the repository"
                    },
                    {
                        "status": "ready",
                        "step": "Git credentials services"
                    },
                    {
                        "status": "ready",
                        "step": "Authentication and proxying services"
                    },
                    {
                        "status": "ready",
                        "step": "Auxiliary session services"
                    },
                    {
                        "status": "ready",
                        "step": "Starting session"
                    }
                ],
                "message": "The resource quota has been exceeded.",
                "readyNumContainers": 7,
                "state": "failed",
                "totalNumContainers": 7,
                "warnings": []
            },
            "url": "https://renku-ci-ui-2942.dev.renku.ch/sessions/flora-2eth-another-2dplayground-2dp-465de908"
        }
    }
```

leafty commented:
Also, it seems that updates to GPU annotations are not taken into account.

olevski commented:
@leafty the action to use a specific amalthea from a specific branch is not published and it was being used only in this PR. If  the deployment from your PR in the ui repo amalthea was not getting deployed. I manually updated the image to test a few times and that is how we got the proper amalthea image which stuck around for a bit. 

But then when we redeployed this got wiped and we reverted to the regular amalthea image. 

I also found the issue when the session is getting stuck on failed and fixed that.

Lastly, I updated the PR description in your PR in the ui repo and in here. So now things should work.

leafty commented:
> @leafty the action to use a specific amalthea from a specific branch is not published and it was being used only in this PR. If the deployment from your PR in the ui repo amalthea was not getting deployed. I manually updated the image to test a few times and that is how we got the proper amalthea image which stuck around for a bit.
> 
> But then when we redeployed this got wiped and we reverted to the regular amalthea image.
> 
> I also found the issue when the session is getting stuck on failed and fixed that.
> 
> Lastly, I updated the PR description in your PR in the ui repo and in here. So now things should work.

The PATCH endpoint seems to work except that GPU annotations are not cleared when needed (change from class with GPU to on with none).

Also, I don't know how feasible it would be to report that the session may be starting soon. On the UI side, the session is shown as failing for one or two minutes before the state goes to "starting". For users, it feels that nothing is happening during that time.

olevski commented:
@leafty thanks for testing all of this out and finding these edge cases. I addressed the problem with the updates to the pod - it was the result of the statefulset not always causing the pod to restart. And the status of the session should now switch much quicker from Failing to Starting when you update the resource class. After the fix when I tested it took just a second or two to go from Failing to Starting.

olevski commented:
>(e.g. selecting another class which cannot be scheduled), it seems that the session takes a while to start (added ~2 minutes).

This is most likely because you cause the session to be scheduled on another node where the image is not cached and has to be downloaded again. Also we have had issues with how quick switch/k8s can remount the drives when the node changes. There is not much that I can do about his unfortunately.

Panaetius reviewed (changes requested):
should there be a test for this?
